### PR TITLE
Separate "not terminated" state in "loading" and "running"

### LINKED ISSUE #12190: PipelinesRegistry reports loading pipelines as running
The PipelinesRegistry defines [running pipelines](https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/pipelines_registry.rb#L230-L232) as:

```ruby
    def running_pipelines
      select_pipelines { |state| !state.terminated? }
    end
```

This causes pipelines in the loading state to appear as running, and in some cases (for example when monitoring is enabled), one can end up seeing an error until the pipeline finishes loading, such as:
```
[2020-08-19T11:20:03,946][ERROR][logstash.inputs.metrics  ] Failed to create monitoring event {:message=>"For path: http_address. Map keys: [:jvm, :stats]", :error=>"LogStash::Instrument::MetricStore::MetricNotFound"}
```

### PR DESCRIPTION
**Pull request by andsel**
## What does this PR do?
Actually the pipelines (from the PipelineRegistry) point of view are in 2 states `terminated` and `running` where `running` is simply obtained as `not terminated`. This drive to a problem that also pipelines in `loading` status are considered as `running` and so are eligible to be queried for metrics while still building its metrics structure, and so generates a Nil access.
This commit separate the "not terminated" state explicitly in "running" and "loading" to avoid some pipelines in loading state to be defined as "running".

## Why is it important/What is the impact to the user?
This PR fixes a bug that manifests on automatic reloading

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [x] run the reproducer

## How to test this PR locally
Use this reproducer https://github.com/elastic/logstash/issues/12190#issuecomment-733070149 and verify that no errors like:
```
[ERROR][logstash.inputs.metrics  ] Failed to create monitoring event {:message=>"undefined method `[]' for nil:NilClass", :error=>"NoMethodError"}
```
are present in logs.

## Related issues

- Closes #12190

~~## Use cases~~

~~## Screenshots~~

~~## Logs~~

### CODE REVIEW

**Comment by andsel** [comment]
Tried with reproducer: https://github.com/elastic/logstash/issues/12190#issuecomment-733070149 and works smoothly

**Comment by jsvd** [comment]
@andsel I can confirm this fixes the issue during reload, but I'm still seeing the issue during startup:

```yml
- pipeline.id: test
  pipeline.workers: 1
  pipeline.batch.size: 1
  config.string: "input { tcp {port => 3333 } } filter { ruby { code => '' init => 'sleep 1111' } sleep { time => 1 } } output { stdout { codec => dots } }"
```

```
[2020-11-26T11:50:04,045][INFO ][logstash.javapipeline    ][.monitoring-logstash] Starting pipeline {:pipeline_id=>".monitoring-logstash", "pipeline.workers"=>1, "pipeline.batch.size"=>2, "pipeline.batch.delay"=>50, "pipeline.max_inflight"=>2, "pipeline.sources"=>["monitoring pipeline"], :thread=>"#<Thread:0x5067a2f1 run>"}
[2020-11-26T11:50:04,678][INFO ][logstash.javapipeline    ][.monitoring-logstash] Pipeline Java execution initialization time {"seconds"=>0.63}
[2020-11-26T11:50:04,701][INFO ][logstash.javapipeline    ][.monitoring-logstash] Pipeline started {"pipeline.id"=>".monitoring-logstash"}
[2020-11-26T11:50:14,752][ERROR][logstash.inputs.metrics  ] Failed to create monitoring event {:message=>"For path: http_address. Map keys: [:stats, :jvm]", :error=>"LogStash::Instrument::MetricStore::MetricNotFound"}
[2020-11-26T11:50:24,792][ERROR][logstash.inputs.metrics  ] Failed to create monitoring event {:message=>"For path: http_address. Map keys: [:stats, :jvm]", :error=>"LogStash::Instrument::MetricStore::MetricNotFound"}
```

And the worst part, that is not related directly to this fix, is that a stuck initializing pipeline can't be interrupted. The errors we see don't inform you that the issue is in a loading pipeline (no mention of the "test" pipeline in the logs), and "fixing" the pipeline configuration does nothing as the converge step is stuck waiting for the pipeline to reload.

**Comment by andsel** [comment]
I think there is 2 problems:
- the `http_address` metric not present when the monitoring pipelines asks for it
- the shutdown hook is not able to tear down the sleeping thread

**`http_address` metric not present**
The bootstrap sequence executed by Agent consists in identify the sequence of actions to apply to the pipelines, executes each action in a separate thread and join on those threads. After the agent joined those thread, that means it completed the converge_state phase, it starts the webserver.
https://github.com/elastic/logstash/blob/6d9ff9c880fca95a87e634fbc883da8ef1b77085/logstash-core/lib/logstash/agent.rb#L114-L118

Now the internal monitoring pipeline is live, because the thread that the agent spawned to execute the create action for that pipeline is terminated, while the other thread for the main pipeline is still running. The live monitoring pipeline ask for `http_address` metric.

https://github.com/elastic/logstash/blob/6d9ff9c880fca95a87e634fbc883da8ef1b77085/x-pack/lib/monitoring/inputs/metrics/stats_event_factory.rb#L128

but that metric is exposed when the webserver is running:
https://github.com/elastic/logstash/blob/6d9ff9c880fca95a87e634fbc883da8ef1b77085/logstash-core/lib/logstash/webserver.rb#L116

The same problem happens when `http.enable: false` and the server doesn't start:
```
[2020-12-02T10:57:59,438][INFO ][logstash.agent           ] Pipelines running {:count=>2, :running_pipelines=>[:test, :".monitoring-logstash"], :non_running_pipelines=>[]}
[2020-12-02T10:57:59,522][INFO ][logstash.agent           ] HTTP API is disabled (`http.enabled=false`); webserver will not be started.
.........[2020-12-02T10:58:09,392][ERROR][logstash.inputs.metrics  ] Failed to create monitoring event {:message=>"For path: http_address. Map keys: [:stats, :os, :jvm]", :error=>"LogStash::Instrument::MetricStore::MetricNotFound"}
```

So the solution here could be to change the monitoring pipeline to ask for that information iff the http server is running, but probably, not providing this information to downstream (ES monitoring) could generate other problems or corrupted state?

**Comment by andsel** [comment]
After a discussion with @jsvd we can follow up this PR with the PRs:
- one to move the webwerver start before the pipelines converge state. This is also sustained by the thought that from central management perspective it needs an API to manage LS before pipelines convergence, since pipeline can goes up and down. As part of this change, the monitoring pipeline shouldn't expose `http_address` if the webserver is not up and running (to check what this could mean in Kibana dashboards, where we show the address)

- improve the converge process, in case we have a pipeline that so long to initialize and if another reload request come for that same pipeline, LS could kill the oldest thread to make going forward the latest request. Motivation, is that if a user does some error in configuring a pipeline and that pipeline take so long, the user in the mean time spot which is the error in his configuration and fixes it; now it should happen a pipeline reload but is blocked from the first, so the net result is that the user changes the pipeline configuration and doesn't see the reflected changes, this is not such a good experience for him.

Further issues to create to track improvements:
- the watcher should also be notified of slow loading pipelines, this make evident in logs that there is a pipeline that has a slow loading phase, perhaps add also the pipeline name.
- expose the status loading/running/terminate through HTTP API, perhaps we could also imagine a backpressure status, but that's part of the running, we need to reflect more on this.

**Review by jsvd — APPROVED** [review]
LGTM with minor nitpicks

File: logstash-core/lib/logstash/pipelines_registry.rb [hunk 1]
```diff
@@ -35,6 +35,19 @@ def terminated?
...
+      @lock.synchronize do
+        # not terminated and not loading
+        @loading.false? && !@pipeline.finished_execution?
+      end
+    end
+
+    def loading?
+      @lock.synchronize do
+        @loading.true?
+      end
```
[the remark this answers is not in the document]
Comment (andsel) [reply logstash-core/lib/logstash/pipelines_registry.rb:48]: I see what you mean, `@loading` is an `AtomicBoolean` so the (read/write) access is safe by type definition, but the state of a `PipelineState` is composed by the tuple `(@pipeline, @loading)` so to access both or only a part of the state I have to synchronize in the instance lock; this is a rule of good concurrency citizens.
Now looking at the code, `@loading.false?`  is used in `terminated?` and `running?` which is a read of an atomic variable, so it's safe and we could remove from locked-set/get. Keeping the lock we are more conservative and cleanly express the access to shared state, WDYT? Is the removal safe and expressive enought?

[the remark this answers is not in the document]
Comment (jsvd) [reply logstash-core/lib/logstash/pipelines_registry.rb:48]: Thanks for the thoughtful answer. I'm ok with keeping the lock here for consistency, as this isn't a critical code path.

**Comment by andsel** [comment]
Jenkins test this please

**Pull request merged**